contract/models/contract.py

```diff
@@ -472,21 +418,15 @@ def _prepare_recurring_invoices_values(self, date_ref=False):
             contract_lines = contract._get_lines_to_invoice(date_ref)
             if not contract_lines:
                 continue
-            invoice_values = contract._prepare_invoice(date_ref)
+            invoice_vals, move_form = contract._prepare_invoice(date_ref)
+            invoice_vals["invoice_line_ids"] = []
             for line in contract_lines:
-                invoice_values.setdefault("invoice_line_ids", [])
-                invoice_line_values = line._prepare_invoice_line(
-                    invoice_values=invoice_values,
-                )
-                if invoice_line_values:
-                    invoice_values["invoice_line_ids"].append(
-                        (0, 0, invoice_line_values)
-                    )
-            invoices_values.append(invoice_values)
+                invoice_line_vals = line._prepare_invoice_line(move_form=move_form)
```

Comment:
Now that v13 has on-memory ORM, would it be a good moment to get rid of these "prepare" methods? They are not so helpful for performance anymore...

Reply:
I'm afraid it's not possible to remove this hook methods if you need to expand this module, due to 3 things:

- Creation is performed split from writes, so you can only optimize second writing and subsequent ones, but the performance will suffer a bit.
- You still need a hook method in which performing that write, and the general one that creates all invoices is not the ideal one.
- And most important, you lose the context of which things is generating which other things, so the hooks are still very valuable for keeping clean code and promote encapsulation.